### DISCUSSION TITLE
How to send a HTTP request header in OTEL signals emitted from instrumented application

### DISCUSSION THREAD

**Discussion by shadow0wolf** [open-telemetry/opentelemetry-java-instrumentation#9767] [Q&A] [answered]
Send a HTTP request header in OTEL signals emitted from instrumented application , this header has dynamic value which is set by application and changes frequently during application lifecycle (unlike the one set with value "otel.exporter.otlp.headers) .
I went through manual instrumentation docs but couldnt find anything.

**Comment by trask** [comment, marked as answer]
hi @shadow0wolf! check out https://github.com/open-telemetry/opentelemetry-java/pull/4630, does that (internal for now) API address what you need?

**Reply by shadow0wolf** [reply]
exactly my use case , i intent to set a header which can change value during application lifetime , this value is supposed to be sent to a OTEL collector (implementing a custom authenticator for validation) , unfortunately this feature seems not to be available for GRPC exporters and it is still internal.

**Reply by trask** [reply]
it may be useful to open an issue in https://github.com/open-telemetry/opentelemetry-java about the stability plan for this feature

is there a reason you need to use the gRPC exporter? (we are curious about this in general, as we have started guiding people towards the HTTP exporter more)

**Reply by shadow0wolf** [reply]
no specific reason for grpc apart from it being available as an option in case required , the problem with this specifically is https://github.com/open-telemetry/opentelemetry-java/pull/4630 , the supplier  function which adds dynamic headers in telemetry requests is only invoked after a 401 response or during the Sender class initialization , i require it be invoked for each request , (the supplier function is required to be in send method of sender and not just on it's constructor)  , @trask

**Reply by trask** [reply]
is this the same as https://github.com/open-telemetry/opentelemetry-java/issues/5959?

**Reply by shadow0wolf** [reply]
@trask , yes
